File: solution/0800-0899/0877.Stone Game/Solution.py
class Solution:
    def stoneGame(self, piles: List[int]) -> bool:
        @cache
        def dfs(i: int, j: int) -> int:
            if i > j:
                return 0
            return max(piles[i] - dfs(i + 1, j), piles[j] - dfs(i, j - 1))

        return dfs(0, len(piles) - 1) > 0

File: solution/0800-0899/0877.Stone Game/Solution.rs
impl Solution {
    pub fn stone_game(piles: Vec<i32>) -> bool {
        let n = piles.len();
        let mut f = vec![vec![0; n]; n];

        fn dfs(i: usize, j: usize, piles: &Vec<i32>, f: &mut Vec<Vec<i32>>) -> i32 {
            if i == j {
                return piles[i]
            }
            if f[i][j] != 0 {
                return f[i][j];
            }

            let res = (piles[i] - dfs(i + 1, j, piles, f))
                .max(piles[j] - dfs(i, j - 1, piles, f));

            f[i][j] = res;
            res
        }

        dfs(0, n - 1, &piles, &mut f) > 0
    }
}
